### ISSUE TITLE
Static variable usage in classes cause problems in some environments

### ISSUE THREAD

**Issue by siltutorials** [erincatto/Box2D#551]
I am using a slightly unusual setup where my application DLL is reloaded at run-time when the project is recompiled without restarting the entire application. This works great as long as I don't store raw functions pointers, don't use malloc/free (I call to the base "platform layer" to allocate deallocate memory), and I don't use static variables unless I'm okay with them resetting every time the application DLL reloads. This causes a problem when trying to use Box2D in my DLL since the static variables in b2BlockAllocator get reset when the DLL reloads and the s_blockSizeLookupInitialized is only checked in the initializer of b2BlockAllocator. This would cause a crash on the first memory allocation after the DLL reloaded for the first time.

This isn't a hard problem to fix as far as I can tell if b2BlockAllocator just checks s_blockSizeLookupInitialized and reinitializes the array in Allocate and Free. However, I'm not too familiar with the inner workings of this class so that might have other unintended consequences.

There was also some static variables in b2Timer that I fixed in a similar manner in my version of the code. Also b2Contact has a static register table which I haven't fixed yet because I'm unsure exactly what it does.

Also as a side note, I have already rerouted b2Alloc and b2Free to my own functions that work without direct calls to malloc/free.

Basically these static variables are causing me problems in the setup that I have and I would like to suggest making some small changes to make these static variable usages more friendly to environments like mine. I understand that this is not a typical environment but static variables don't seem to be too widespread in Box2D and they don't seem to be hard to fix in the 3 cases that I could find. Let me know if you want any more information about my setup or the symptoms I was experiencing

**Comment by siltutorials** [comment]
Turns out that I have a whole load of other problems with the virtual function table allocated by C++ for overloaded functions so remedying the static variable usage doesn't completely solve the problem. I suppose will close this issue.
